docs/lazy-adr/adr-004-mvp-light-client.md

```diff
@@ -0,0 +1,292 @@
++++ b/cmd/tendermint/commands/light.go	(date 1620546761822)
+@@ -64,6 +64,8 @@
+ 	dir                string
+ 	maxOpenConnections int
+
++	daSampling     bool
++	numSamples     uint32
+ 	sequential     bool
+ 	trustingPeriod time.Duration
+ 	trustedHeight  int64
+@@ -101,6 +103,10 @@
+ 	LightCmd.Flags().BoolVar(&sequential, "sequential", false,
+ 		"sequential verification. Verify all headers sequentially as opposed to using skipping verification",
+ 	)
++	LightCmd.Flags().BoolVar(&daSampling, "da-sampling", false,
++		"data availability sampling. Verify each header (sequential verification), additionally verify data availability via data availability sampling",
++	)
++	LightCmd.Flags().Uint32Var(&numSamples, "num-samples", 15, "Number of data availability samples until block data deemed available.")
+ }
+```
+
+For the Data Availability sampling, the light client will have to run an IPFS node.
+It makes sense to make this mostly opaque to the user as everything around IPFS can be [configured](https://github.com/ipfs/go-ipfs/blob/d6322f485af222e319c893eeac51c44a9859e901/docs/config.md) in the `$IPFS_PATH`.
+This IPFS path should simply be a sub-directory inside the light client's [directory](https://github.com/lazyledger/lazyledger-core/blob/cbf1f1a4a0472373289a9834b0d33e0918237b7f/cmd/tendermint/commands/light.go#L86-L87).
+We can later add the ability to let users configure the IPFS setup more granular.
+
+**Note:** DAS should only be compatible to sequential verification.
+In case a light client is parametrized to run DAS and skipping verification, the CLI should return an easy-to-understand warning or even an error explaining why this does not make sense.
+
+### Light Client Protocol with DAS
+
+#### Light Store
+
+The light client stores data in its own [badgerdb instance](https://github.com/lazyledger/lazyledger-core/blob/50f722a510dd2ba8e3d31931c9d83132d6318d4b/cmd/tendermint/commands/light.go#L125) in the given directory:
+
+```go
+db, err := badgerdb.NewDB("light-client-db", dir)
+```
+
+While it is not critical for this feature, we should at least try to re-use that same DB instance for the local ipld store.
+Otherwise, we introduce yet another DB instance; something we want to avoid, especially on the long run (see [#283](https://github.com/lazyledger/lazyledger-core/issues/283)).
+For the first implementation, it might still be simpler to create a separate DB instance and tackle cleaning this up in a separate pull request, e.g. together with other [instances]([#283](https://github.com/lazyledger/lazyledger-core/issues/283)).
+
+#### RPC
+
+No changes to the RPC endpoints are absolutely required.
+Although, for convenience and ease of use, we should either add the `DAHeader` to the existing [Commit](https://github.com/lazyledger/lazyledger-core/blob/cbf1f1a4a0472373289a9834b0d33e0918237b7f/rpc/core/routes.go#L25) endpoint, or, introduce a new endpoint to retrieve the `DAHeader` on demand and for a certain height or block hash.
```

Comment:
> It will be included in other data types soon.

None of the existing RPC responses a tendermint light client requests will contain the DAHeader though (the BlockID in #312 is only temporary.

> In the end, we will rely mostly on Header.DataHash instead.

True! 

> Redundantly requesting it separately is obviously not a good idea from long-living subscriptions POV, though subscribing for the whole EventNewBlock is also a bloated approach in some use-cases and EventNewBlockHeader does not contain DAHeader.

The light client itself doesn't use the event system. Are you talking about the future where the light clients will subscribe to headers as you described in https://github.com/lazyledger/lazyledger-core/issues/86#issuecomment-831182564 ?

Reply:
> Are you talking about the future where the light clients will subscribe to headers as you described in #86 (comment) ?

Nope. I was talking about EventBus within ll-core. The light client can subscribe to new things through RPC with EventBus as a backbone. 

From my understanding, EventNewBlock and EventNewBlockHeader are the most used RPC endpoint/subscription for the light clients to be always on track with the network, and doing separate requests for DAHeader to perform DAS on every subscription message does not make sense for me.

Reply:
@liamsi, so you didn't rely on DAHeader in BlockID as temporary, but you did rely on DAHeader instead of DataHash because it weeks away? 🤔  Kinda controversial 😄, but I understand that it was not clear before that DataHash can be dropped in instead of DAHeader so I am ok with that.

Reply:
> @liamsi, so you didn't rely on DAHeader in BlockID as temporary, but you did rely on DAHeader instead of DataHash because it weeks away? 🤔 Kinda controversial 😄, 

What exactly is controversial about it and what are you suggesting to do about it? Please, always keep include constructive suggestions when you criticize something. Otherwise, it is not clear what improvements could be made here.

Just to clarify: Everything in this ADR is written in a way that a) does not depend on any feature that is not clearly defined or implemented yet and b) does not introduce any technical debt. 

I don't really see why it is harmful but we can (very) easily remove the additional rpc end-point in the near future if we wanted to.

It is clear that the light client could simply do with the data hash but: 1) how to do this via IPFS is not defined nor implemented yet and 2) we wouldn't have been able to ship a DAS light client as quickly if we waited on that.

> The light client can subscribe to new things through RPC with EventBus as a backbone.

That is a valuable and cool observation but it's still not how the current default light client operates: basically the rpc endpoint of the light client "drives" the light client (it gets updated on demand if you e.g. request anything with a newer height from the light client's RPC).

That said, for future light client implementations, we could use the event bus. If you want, you can write an ADR for that (it's not urgent but we should write an ADR for different variants for light clients we want, another one is #86 for instance).

Does this make sense?